lib/event_summarizer/idv_matcher.rb

```diff
@@ -0,0 +1,538 @@
+        gpo? && !self.significant_events.any? { |e| e.type == :gpo_code_success }
+      end
+
+      def ipp?
+        self.significant_events.any? { |e| e.type == :start_ipp }
+      end
+
+      def ipp_pending?
+        ipp? && !self.significant_events.any? { |e| e.type == :ipp_enrollment_complete }
+      end
+
+      def successful?
+        self.significant_events.any? do |e|
+          e.type == :verified
+        end
+      end
+
+      def workflow_complete?
+        gpo? || ipp? || flagged_for_fraud? || successful?
+      end
+    end.freeze
+
+    SignificantIdvEvent = Data.define(
+      :timestamp,
+      :type,
+      :description,
+    ).freeze
+
+    attr_reader :current_idv_attempt
+    attr_reader :idv_attempts
+    attr_reader :idv_abandoned_event
+
+    def initialize
+      @idv_attempts = []
+      @current_idv_attempt = nil
+    end
+
+    # @return {Hash,nil}
+    def handle_cloudwatch_event(event)
+      case event['name']
+        when IDV_WELCOME_SUBMITTED_EVENT
+          start_new_idv_attempt(event:)
+
+        when IDV_FINAL_RESOLUTION_EVENT
+
+          for_current_idv_attempt(event:) do
+            handle_final_resolution_event(event:)
+          end
+
+        when IDV_GPO_CODE_SUBMITTED_EVENT
+          for_current_idv_attempt(event:) do
+            handle_gpo_code_submission(event:)
+          end
+
+        when IPP_ENROLLMENT_STATUS_UPDATED_EVENT
+          for_current_idv_attempt(event:) do
+            handle_ipp_enrollment_status_update(event:)
+          end
+
+        when IDV_IMAGE_UPLOAD_VENDOR_SUBMITTED_EVENT
+          for_current_idv_attempt(event:) do
+            handle_image_upload_vendor_submitted(event:)
+          end
+
+        when IDV_VERIFY_PROOFING_RESULTS_EVENT
+          for_current_idv_attempt(event:) do
+            handle_verify_proofing_results_event(event:)
+          end
+
+        when PROFILE_ENCRYPTION_INVALID_EVENT
+          for_current_idv_attempt(event:) do
+            handle_profile_encryption_error(event:)
+          end
+
+        when RATE_LIMIT_REACHED_EVENT
+          handle_rate_limit_reached(event:)
+
+        else
+          if ENV['LOG_UNHANDLED_EVENTS']
+            warn "#{event['@timestamp']} #{event['name']}"
+          end
+      end
+
+      check_for_idv_abandonment(event)
+    end
+
+    def finish
+      finish_current_idv_attempt
+
+      self.idv_attempts.map { |a| summarize_idv_attempt(a) }.tap do
+        idv_attempts.clear
+      end
+    end
+
+    private
+
+    def add_significant_event(
+      timestamp:,
+      type:,
+      description:
+    )
+      current_idv_attempt.significant_events << SignificantIdvEvent.new(
+        timestamp:,
+        type:,
+        description:,
+      )
+    end
+
+    def check_for_idv_abandonment(event)
+      return if current_idv_attempt.nil?
+
+      looks_like_idv = /^idv/i.match(event['name'])
+      return if !looks_like_idv
+
+      if idv_abandoned_event.nil?
+        @idv_abandoned_event = event
```

Comment:
does this work on an instance of `Data`? I thought this fails because each instance is frozen